bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/api/impl/EcovacsApiUrlFactory.java

```diff
@@ -68,6 +74,11 @@ private static String getMainUrl(EcovacsApiConfiguration config) {
                 config.getAppCode(), config.getAppVersion(), config.getChannel(), config.getDeviceType());
     }
 
+    private static String getAppUrl(EcovacsApiConfiguration config) {
+        String continentSuffix = "cn".equalsIgnoreCase(config.getCountry()) ? "" : "-" + config.getContinent();
+        return String.format("https://portal%1$s.ecouser.net/app", continentSuffix);
+    }
```

Comment:
Perhaps the duplication from `getPortalUrl` could be reduced? For example by extracting creation of `continentSuffix` to a separate method or even creation of the root path shared between portal and app.

Reply:
Did the latter (and renamed the app endpoint DTO, because the server is still `portal.ecovacs.net`).